src/adaptix/_internal/morphing/generic_provider.py

```diff
@@ -407,9 +410,21 @@ def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
         if all(dumper == as_is_stub for dumper in dumpers):
             return as_is_stub
 
-        dumper_type_dispatcher = ClassDispatcher(
-            {type(None) if case.origin is None else case.origin: dumper for case, dumper in zip(norm.args, dumpers)},
-        )
+        case_dumpers = {}
+        literal_type, literal_dumper = None, None
+
+        for case, dumper in zip(norm.args, dumpers):
+            if case.origin is Literal:
+                literal_type, literal_dumper = case, dumper
+            key = type(None) if case.origin is None else case.origin
+            case_dumpers[key] = dumper
+
+        dumper_type_dispatcher = ClassDispatcher(case_dumpers)
+
+        if literal_type:
+            literal_cases = literal_type.args
+            return self._get_dumper_for_literal(dumper_type_dispatcher, literal_dumper, literal_cases)
```

Comment:
I think that this code will be more clear if there will be separate function handing Literal case.
Such function should return ``Optional[Dumper]``

Reply:
done